File: client-mobile/modules/message-selection/android/src/main/java/cloud/baizhi/messageselection/MessageSelectionModule.kt
package cloud.baizhi.messageselection

import android.os.Bundle
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class MessageSelectionModule : Module(), MessageSelectionEventListener {
  override fun definition() = ModuleDefinition {
    Name("MagicChatMessageSelection")

    Events("onMessageAction")

    OnCreate {
      MessageSelectionEvents.addListener(this@MessageSelectionModule)
    }

    OnDestroy {
      MessageSelectionEvents.removeListener(this@MessageSelectionModule)
    }
  }

  override fun onMessageSelectionAction(action: String, messageId: String) {
    sendEvent(
      "onMessageAction",
      Bundle().apply {
        putString("action", action)
        putString("messageId", messageId)
      }
    )
  }
}

File: client-mobile/modules/message-selection/ios/MessageSelectionModule.swift
import ExpoModulesCore
import Foundation

private let messageSelectionNotification = Notification.Name(
  "MagicChatMessageSelectionAction"
)

public final class MessageSelectionModule: Module {
  private var observer: NSObjectProtocol?

  public func definition() -> ModuleDefinition {
    Name("MagicChatMessageSelection")

    Events("onMessageAction")

    OnCreate {
      self.observer = NotificationCenter.default.addObserver(
        forName: messageSelectionNotification,
        object: nil,
        queue: .main
      ) { [weak self] notification in
        guard
          let action = notification.userInfo?["action"] as? String,
          let messageId = notification.userInfo?["messageId"] as? String
        else {
          return
        }

        self?.sendEvent("onMessageAction", [
          "action": action,
          "messageId": messageId,
        ])
      }
    }

    OnDestroy {
      if let observer = self.observer {
        NotificationCenter.default.removeObserver(observer)
        self.observer = nil
      }
    }
  }
}
